word_segmentation.py

```diff
@@ -0,0 +1,709 @@
+            bies[3, i] = 1
+            word_ind += 1
+            continue
+    return bies
+
+
+def get_bies_string_from_softmax (mat):
+    """
+    Computes estimated BIES based on a softmax matrix of size n*4.
+    Each row of matrix gives four floats that add up to 1, which are probability of B, I, E,  and S.
+    This function simply pick the one with highest probability. In ties, it picks B over I, I over E, and E over S.
+    mat: Input matrix that contains softmax probabilities. Dimension is n*4 where n is length of the string.
+    """
+    out = ""
+    for i in range(mat.shape[0]):
+        max_softmax = max(mat[i, :])
+        if mat[i, 0] == max_softmax:
+            out += "b"
+        elif mat[i, 1] == max_softmax:
+            out += "i"
+        elif mat[i, 2] == max_softmax:
+            out += "e"
+        elif mat[i, 3] == max_softmax:
+            out += "s"
+    return out
+
+
+def remove_tags(line, st_tag, fn_tag):
+    """
+    Given a string and two substrings, remove any text between these tags.
+    line: the input string
+    st_tag: the first substring
+    fn_tag: the secibd substring
+    It handles spaces around tags as follows:
+        abc|<NE>def</NE>|ghi      ---> abc|ghi
+        abc| |<NE>def</NE>|ghi    ---> abc| |ghi
+        abc|<NE>def</NE>| |ghi    ---> abc| |ghi
+        abc| |<NE>def</NE>| |ghi  ---> abc| |ghi
+    """
+
+    new_line = ""
+    st_ind = 0
+    while st_ind < len(line):
+        if line[st_ind: st_ind+len(st_tag)] == st_tag:
+            fn_ind = st_ind
+            while fn_ind < len(line):
+                if line[fn_ind: fn_ind+len(fn_tag)] == fn_tag:
+                    fn_ind = fn_ind+len(fn_tag) + 1
+                    if st_ind -2 >= 0 and fn_ind+2 <= len(line):
+                        if line[st_ind-2:st_ind] == " |" and line[fn_ind:fn_ind+2] == " |":
+                            fn_ind += 2
+                    st_ind = fn_ind
+                    break
+                else:
+                    fn_ind += 1
+        if st_ind < len(line):
+            new_line += line[st_ind]
+        st_ind += 1
+    return new_line
+
+
+def clean_line(line):
+    """
+    This line cleans a line as follows such that it is ready for process by different components of the code. It returns
+    the clean line or -1, if the line should be omitted.
+        1) remove tags and https from the line.
+        2) Put a | at the begining and end of the line if it isn't already there
+        3) if line is very short (len < 3) or if it is all in English or it has a link in it, return -1
+    """
+
+    # Remove lines with links
+    if "http" in line or len(line) == 0:
+        return -1
+
+    # Remove texts between following tags
+    line = remove_tags(line, "<NE>", "</NE>")
+    line = remove_tags(line, "<AB>", "</AB>")
+
+    # Remove lines that are all fully in English
+    if is_all_english(line):
+        return -1
+
+    # Exclude cases such as "", " ", "| ", " |", etc.
```

Comment:
Doesn't `is_all_english` catch those cases?

Reply:
In my original `is_all_english` I excluded spaces and bars before checking if a text is in ASCII. I don't remember why I did that, but I remember I had a reason. For now, I rewrite the `is_all_englihs` function and remove these special cases, hoping it doesn't result in a bug (I didn't detect any by running some of the main functions).